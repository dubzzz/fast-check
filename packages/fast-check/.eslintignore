bundle/
node_modules/
lib/
lib-*/
perf/
jest*.js
jest*.cjs