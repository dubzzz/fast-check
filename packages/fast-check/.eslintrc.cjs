module.exports = {
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/eslint-recommended',
    'plugin:@typescript-eslint/recommended',
    'prettier',
  ],
  rules: {
    '@typescript-eslint/explicit-module-boundary-types': 'error',
    '@typescript-eslint/explicit-function-return-type': 'off',
    '@typescript-eslint/interface-name-prefix': 'off',
    '@typescript-eslint/no-empty-function': 'warn',
    '@typescript-eslint/no-empty-interface': 'off',
    '@typescript-eslint/no-explicit-any': 'off',
    '@typescript-eslint/no-this-alias': 'warn',
    '@typescript-eslint/no-unused-vars': ['error', { argsIgnorePattern: '^_' }],
    '@typescript-eslint/no-use-before-define': 'warn',
  },
  overrides: [
    {
      files: ['*.spec.ts', 'test/unit/**/*.ts', 'test/e2e/**/*.ts'],
      rules: {
        '@typescript-eslint/no-empty-function': 'off',
        '@typescript-eslint/no-use-before-define': 'off',
        '@typescript-eslint/no-non-null-assertion': 'off',
      },
    },
  ],
};
