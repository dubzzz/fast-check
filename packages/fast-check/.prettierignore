bundle/
coverage/
docs/
documentation/Arbitraries.md
lib/
lib-*/