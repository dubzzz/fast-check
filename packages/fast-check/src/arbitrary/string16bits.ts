import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import type { ArrayConstraintsInternal } from './array';
import { array } from './array';
import { char16bits } from './char16bits';
import type { StringSharedConstraints } from './_shared/StringSharedConstraints';
import { charsToStringMapper, charsToStringUnmapper } from './_internals/mappers/CharsToString';
import { createSlicesForString } from './_internals/helpers/SlicesForStringBuilder';
export type { StringSharedConstraints } from './_shared/StringSharedConstraints';

const safeObjectAssign = Object.assign;

/**
 * For strings of {@link char16bits}
 *
 * @param constraints - Constraints to apply when building instances (since 2.4.0)
 *
 * @remarks Since 0.0.11
 * @public
 */
export function string16bits(constraints: StringSharedConstraints = {}): Arbitrary<string> {
  const charArbitrary = char16bits();
  const experimentalCustomSlices = createSlicesForString(charArbitrary, charsToStringUnmapper);
  // TODO - Move back to object spreading as soon as we bump support from es2017 to es2018+
  const enrichedConstraints: ArrayConstraintsInternal<string> = safeObjectAssign(safeObjectAssign({}, constraints), {
    experimentalCustomSlices,
  });
  return array(charArbitrary, enrichedConstraints).map(charsToStringMapper, charsToStringUnmapper);
}
