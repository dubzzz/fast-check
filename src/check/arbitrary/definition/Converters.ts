import { Arbitrary, assertIsArbitrary } from './Arbitrary';
import { ArbitraryWithContextualShrink } from './ArbitraryWithContextualShrink';
import { ConverterFromNext } from './ConverterFromNext';
import { ConverterToNext } from './ConverterToNext';
import { assertIsNextArbitrary, NextArbitrary } from './NextArbitrary';

/**
 * Convert an instance of NextArbitrary to an instance of Arbitrary
 * @param arb - The instance to be converted
 * @public
 */
export function convertFromNext<T>(arb: NextArbitrary<T>): Arbitrary<T> {
  if (ConverterToNext.isConverterToNext(arb)) {
    return arb.arb;
  }
  assertIsNextArbitrary(arb);
  return new ConverterFromNext(arb);
}

/**
 * Convert an instance of NextArbitrary to an instance of ArbitraryWithContextualShrink
 * @param arb - The instance to be converted
 * @param legacyShrunkOnceContext - Default context to be returned when shrunk once
 * @public
 */
export function convertFromNextWithShrunkOnce<T>(
  arb: NextArbitrary<T>,
  legacyShrunkOnceContext: unknown
): ArbitraryWithContextualShrink<T> {
  if (ConverterToNext.isConverterToNext(arb)) {
    if (
      !('contextualShrink' in arb.arb) ||
      !('contextualShrinkableFor' in arb.arb) ||
      !('shrunkOnceContext' in arb.arb) ||
      !('shrink' in arb.arb) ||
      !('shrinkableFor' in arb.arb)
    ) {
      throw new Error('Conversion rejected: Underlying arbitrary is not compatible with ArbitraryWithContextualShrink');
    }
    return arb.arb;
  }
  assertIsNextArbitrary(arb);
  return new ConverterFromNext(arb, legacyShrunkOnceContext);
}

/**
 * Convert an instance of Arbitrary to an instance of NextArbitrary
 * @param arb - The instance to be converted
 * @public
 */
export function convertToNext<T>(arb: Arbitrary<T>): NextArbitrary<T> {
  if (ConverterFromNext.isConverterFromNext(arb)) {
    return arb.arb as NextArbitrary<T>;
  }
  assertIsArbitrary(arb);
  return new ConverterToNext(arb);
}
