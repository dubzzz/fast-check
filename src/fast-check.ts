import * as fc from './fast-check-default';
export default fc;

export * from './fast-check-default';
