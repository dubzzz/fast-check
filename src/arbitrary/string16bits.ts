import { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { array } from './array';
import { char16bits } from './char16bits';
import { StringSharedConstraints } from './_shared/StringSharedConstraints';
import { charsToStringMapper, charsToStringUnmapper } from './_internals/mappers/CharsToString';
export { StringSharedConstraints } from './_shared/StringSharedConstraints';

/**
 * For strings of {@link char16bits}
 *
 * @param constraints - Constraints to apply when building instances (since 2.4.0)
 *
 * @remarks Since 0.0.11
 * @public
 */
export function string16bits(constraints: StringSharedConstraints = {}): Arbitrary<string> {
  return array(char16bits(), constraints).map(charsToStringMapper, charsToStringUnmapper);
}
