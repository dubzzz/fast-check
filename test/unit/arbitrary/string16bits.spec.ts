import * as fc from '../../../lib/fast-check';
import { string16bits } from '../../../src/arbitrary/string16bits';

import { convertToNext } from '../../../src/check/arbitrary/definition/Converters';
import {
  assertGenerateProducesSameValueGivenSameSeed,
  assertGenerateProducesCorrectValues,
  assertGenerateProducesValuesFlaggedAsCanGenerate,
  assertShrinkProducesSameValueWithoutInitialContext,
  assertShrinkProducesCorrectValues,
  assertShrinkProducesValuesFlaggedAsCanGenerate,
} from '../check/arbitrary/generic/NextArbitraryAssertions';

describe('string16bits (integration)', () => {
  type Extra = { minLength?: number; maxLength?: number };
  const extraParameters: fc.Arbitrary<Extra> = fc
    .tuple(fc.nat({ max: 30 }), fc.nat({ max: 30 }), fc.boolean(), fc.boolean())
    .map(([min, gap, withMin, withMax]) => ({
      minLength: withMin ? min : undefined,
      maxLength: withMax ? min + gap : undefined,
    }));

  const isCorrect = (value: string, extra: Extra) => {
    if (extra.minLength !== undefined) {
      expect(value.length).toBeGreaterThanOrEqual(extra.minLength);
    }
    if (extra.maxLength !== undefined) {
      expect(value.length).toBeLessThanOrEqual(extra.maxLength);
    }
  };

  const string16bitsBuilder = (extra: Extra) => convertToNext(string16bits(extra));

  it('should generate the same values given the same seed', () => {
    assertGenerateProducesSameValueGivenSameSeed(string16bitsBuilder, { extraParameters });
  });

  it('should only generate correct values', () => {
    assertGenerateProducesCorrectValues(string16bitsBuilder, isCorrect, { extraParameters });
  });

  it('should recognize values that would have been generated using it during generate', () => {
    assertGenerateProducesValuesFlaggedAsCanGenerate(string16bitsBuilder, { extraParameters });
  });

  it('should shrink towards the same values given the same seed', () => {
    assertGenerateProducesSameValueGivenSameSeed(string16bitsBuilder, { extraParameters });
  });

  it('should be able to shrink without any context', () => {
    assertShrinkProducesSameValueWithoutInitialContext(string16bitsBuilder, { extraParameters });
  });

  it('should only shrink towards correct values', () => {
    assertShrinkProducesCorrectValues(string16bitsBuilder, isCorrect, { extraParameters });
  });

  it('should recognize values that would have been generated using it during shrink', () => {
    assertShrinkProducesValuesFlaggedAsCanGenerate(string16bitsBuilder, { extraParameters });
  });
});
