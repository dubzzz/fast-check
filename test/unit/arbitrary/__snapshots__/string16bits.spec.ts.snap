// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`string16bits (integration) should be able to shrink Hey 🐱! 1`] = `
"\\"Hey 🐱!\\"
├> \\"\\"
├> \\" 🐱!\\"
|  ├> \\"\\\\udc31!\\"
|  |  ├> \\"!\\"
|  |  |  ├> \\"\\"
|  |  |  └> \\" \\"
|  |  |     └> \\"\\"
|  |  ├> \\" !\\"
|  |  |  ├> \\"!\\"
|  |  |  |  ├> \\"\\"
|  |  |  |  └> \\" \\"
|  |  |  |     └> \\"\\"
|  |  |  ├> \\" \\"
|  |  |  |  └> \\"\\"
|  |  |  └> \\"  \\"
|  |  |     ├> \\" \\"
|  |  |     |  └> \\"\\"
|  |  |     └> \\" \\"
|  |  |        └> \\"\\"
|  |  ├> \\"渙!\\"
|  |  |  ├> \\"!\\"
|  |  |  |  ├> \\"\\"
|  |  |  |  └> \\" \\"
|  |  |  |     └> \\"\\"
|  |  |  ├> \\"㜍!\\"
|  |  |  |  ├> \\"!\\"
|  |  |  |  |  ├> \\"\\"
|  |  |  |  |  └> \\" \\"
|  |  |  |  |     └> \\"\\"
|  |  |  |  ├> \\"ᮇ!\\"
|  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  ├> \\"හ!\\"
|  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  ├> \\"ۢ!\\"
|  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  ├> \\"ͱ!\\"
|  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  ├> \\"ƹ!\\"
|  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  ├> \\"Ý!\\"
|  |  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  ├> \\"\\\\u0010!\\"
|  |  |  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  ├> \\"X!\\"
|  |  |  |  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  ├> \\"<!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\".!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"'!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"$!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\\\\\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"!!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\" !\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\"  \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     ├> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |        └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\"! \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     ├> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     ├> \\"  \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  ├> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  |  └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |        ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |        └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |           └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\\\\\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |        └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\"\\\\\\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     ├> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     ├> \\"! \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  ├> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  |  └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  ├> \\"  \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  |  ├> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  |  |  └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  └> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |     ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |     └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |        └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"\\\\\\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |        ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |        └> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |           ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |           └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |              └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"#!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\\\\\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\" !\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\"  \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     ├> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |        └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"!!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\" !\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\"  \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     ├> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |        └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\"! \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     ├> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     ├> \\"  \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  ├> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  |  └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |        ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |        └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |           └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\\\\\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |        └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\"\\\\\\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     ├> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     ├> \\"  \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  ├> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  |  └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     ├> \\"! \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  ├> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  |  └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  ├> \\"  \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  |  ├> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  |  |  └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |  └> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |     ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |     └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     |        └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> \\"\\\\\\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |        ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |        ├> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |        |  └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |        └> \\"!\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |           ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |           └> \\" \\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |              └> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"#\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> \\"\\\\\\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> …
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  └> …
|  |  |  |  |  |  └> …
|  |  |  |  |  └> …
|  |  |  |  └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;
